urllib3/util/hpkp.py

```diff
@@ -0,0 +1,269 @@
+        self.db.store_host(pin)
+
+        # TODO: Check if the RFC wants us to immediately validate the
+        # connection: if it does, do it.
+        return
+
+    def _find_valid_pinned_host(self, domain, hosts):
+        """
+        Locates the most specific valid ``KnownPinnedHost`` for a given domain.
+
+        If none of the owned hosts are valid for this domain, returns ``None``.
+
+        :param domain: The domain being validated.
+        :param hosts: A list of ``KnownPinnedHost`` objects whose domains are
+            either the same as ``domain``, or are parent domains of ``domain``.
+        """
+        for host in hosts:
+            # First, check whether this host is still valid. If it's not, throw
+            # it away and move on.
+            if (host.start_date + host.max_age) > time.time():
+                self.db.invalidate_host(host.domain)
+                continue
+
+            # Now, check whether this KnownPinnedHost applies to this
+            # connection. It does if the domain either exactly matches the KPH
+            # domain, or if the KPH has include_subdomains set to True.
+            if domain == host.domain or host.include_subdomains:
+                return host
+
+
+# An object that holds information about a KnownPinnedHost. This object
+# basically just stores string data, so there's no real need to do anything
+# clever here.
+KnownPinnedHost = collections.namedtuple(
+    'KnownPinnedHost',
+    ['domain', 'pins', 'max_age', 'include_subdomains', 'report_uri', 'start_date']
+)
+
+
+def parse_public_key_pins(header, domain):
+    """
+    Parses a Public-Key-Pins header, returning a KnownPinnedHost. Invalid
+    headers cause exceptions to be thrown.
+
+    :param header: The PKP header value.
+    :param domain: The domain name used in the request. Must not be an IP
+        address.
+    """
+    # TODO: This method is super complicated, refactor.
+    # Parse the directives. Split any OWS, and then split the directives.
+    directives = header.split(';')
+
+    pin_directives = []
+    max_age = None
+    include_subdomains = False
+    report_uri = None
+
+    for directive in directives:
+        # Strip any OWS, then split into name and value. All directives have
+        # names, but frustratingly they don't all have values.
+        directive = directive.strip().split('=', 1)
+        name = directive[0]
+        try:
+            value = directive[1]
+        except IndexError:
+            value = None
+
+        if name == 'pin-sha256':
+            value = unquote_string(value)
+            pin_directives.append(value)
+        elif name == 'max-age':
+            # There may only be one max-age directive. Be careful about
+            # policing this, we don't want to accidentally allow more than one.
+            if max_age:
+                raise HPKPError("Multiple max-age directives in PKP header")
+
+            max_age = int(unquote_string(value))
```

Comment:
What are your thoughts on the RFC's suggestions for a "maximum" max-age: https://tools.ietf.org/html/rfc7469#section-4.1 ?